extensions/amp-a4a/0.1/refresh-manager.js

```diff
@@ -0,0 +1,366 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {isExperimentOn} from '../../../src/experiments';
+import {Services} from '../../../src/services';
+import {dev, user} from '../../../src/log';
+import {IntersectionObserverPolyfill} from '../../../src/intersection-observer-polyfill'; // eslint-disable-line max-len
+
+/**
+ * - visibilePercentageMin: The percentage of pixels that need to be on screen
+ *   for the creative to be considered "visible".
+ * - continuousTimeMin: The amount of continuous time, in milliseconds, that
+ *   the creative must be on screen for in oreder to be considered "visible".
+ *
+ * @typedef {{
+ *   visiblePercentageMin: number,
+ *   continuousTimeMin: number,
+ * }}
+ */
+export let RefreshConfig;
+
+export const MIN_REFRESH_INTERVAL = 30;
+export const DATA_ATTR_NAME = 'data-enable-refresh';
+export const DATA_MANAGER_ID_NAME = 'data-amp-ad-refresh-id';
+export const METATAG_NAME = 'amp-ad-enable-refresh';
+
+const TAG = 'AMP-AD';
+
+/**
+ * Defines the DFA states for the refresh cycle.
+ *
+ * 1. All newly registered elements begin in the INITIAL state.
+ * 2. Only when the element enters the viewport with the specified
+ *    intersection ratio does it transition into the VIEW_PENDING state.
+ * 3. If the element remains in the viewport for the specified duration, it
+ *    will then transition into the REFRESH_PENDING state, otherwise it will
+ *    transition back into the INITIAL state.
+ * 4. The element will remain in REFRESH_PENDING state until the refresh
```

Comment:
Oh, sorry, you're correct. We impose an artificial 1s delay so that the ad loader can be seen by the user before transitioning from one creative to another.